docs/core/whats-new/dotnet-9.md

```diff
@@ -0,0 +1,84 @@
+
+This article has been updated for .NET 9 Preview 1.
+
+## .NET library updates
+
+- [Serialization](#serialization)
+- [LINQ](#linq)
+- [Collections](#collections)
+- [Cryptography](#cryptography)
+
+### Serialization
+
+### LINQ
+
+New methods `CountBy` and `AggregateBy` have been introduced. These methods make it possible to aggregate state by key without needing to allocate intermediate groupings via <xref:System.Linq.Enumerable.GroupBy%2A>.
+
+`CountBy` lets you quickly calculate the frequency of each key. The following example finds the word that occurs most frequently in a text string.
+
+:::code language="csharp" source="snippets/dotnet-9/csharp/Linq.cs" id="CountBy":::
+
+`AggregateBy` lets you implement more general-purpose workflows. The following example shows how you can calculate scores that are associated with a given key.
+
+:::code language="csharp" source="snippets/dotnet-9/csharp/Linq.cs" id="AggregateBy":::
+
+### Collections
+
+#### PriorityQueue.Remove() method
+
+.NET 6 introduced the <xref:System.Collections.Generic.PriorityQueue%602> collection, which provides a simple and fast array-heap implementation. One issue with array heaps in general is that they [don't support priority updates](https://github.com/dotnet/runtime/issues/44871), which makes them prohibitive for use in algorithms such as variations of [Dijkstra's algorithm](https://en.wikipedia.org/wiki/Dijkstra%27s_algorithm#Using_a_priority_queue).
+
+While it's not possible to implement efficient `O(log n)` priority updates in the existing collection, the new `PriorityQueue.Remove` method makes it possible to emulate priority updates (albeit at `O(n)` time):
+
+:::code language="csharp" source="snippets/dotnet-9/csharp/Collections.cs" id="UpdatePriority":::
+
+This method unblocks users who want to implement graph algorithms in contexts where asymptotic performance isn't a blocker. (Such contexts include education and prototyping.) For example, here's a [toy implementation of Dijkstra's algorithm](https://github.com/dotnet/runtime/blob/16cb41496d595e2568574cfe11c763d5e05136c9/src/libraries/System.Collections/tests/Generic/PriorityQueue/PriorityQueue.Tests.Dijkstra.cs#L46-L76) that uses the new API.
+
+### Cryptography
+
+#### CryptographicOperations.HashData() method
+
+Since .NET 5, some static ["one-shot"](../../standard/security/cryptography-model.md#one-shot-apis) implementations of hash functions and related functions have been implemented. These APIs include <xref:System.Security.Cryptography.SHA256.HashData%2A?displayProperty=nameWithType> and <xref:System.Security.Cryptography.HMACSHA256.HashData%2A?displayProperty=nameWithType>. One-shot APIs are preferable to use because they can provide the best possible performance and reduce or eliminate allocations.
```

Comment:
```suggestion
.NET includes several static ["one-shot"](../../standard/security/cryptography-model.md#one-shot-apis) implementations of hash functions and related functions. These APIs include <xref:System.Security.Cryptography.SHA256.HashData%2A?displayProperty=nameWithType> and <xref:System.Security.Cryptography.HMACSHA256.HashData%2A?displayProperty=nameWithType>. One-shot APIs are preferable to use because they can provide the best possible performance and reduce or eliminate allocations.
```